In [4]:
import pandas as pd

In [12]:
turbine_14 = pd.read_csv("../data_pack/turbine_telemetry_14day_sample.csv")
turbine_14.head(10)

,timestamp,turbine_id,wind_speed,power_kw,availability
0,2026-03-01 00:00:00,T01,7.11,198.3,1.0
1,2026-03-01 00:00:00,T02,5.42,25.6,1.0
2,2026-03-01 00:00:00,T03,8.31,393.2,1.0
3,2026-03-01 00:00:00,T04,4.82,11.3,1.0
4,2026-03-01 00:00:00,T05,4.24,27.3,1.0
5,2026-03-01 01:00:00,T01,5.01,0.0,1.0
6,2026-03-01 01:00:00,T02,6.23,69.1,1.0
7,2026-03-01 01:00:00,T03,5.41,41.0,1.0
8,2026-03-01 01:00:00,T04,4.46,0.0,1.0
9,2026-03-01 01:00:00,T05,8.12,410.7,1.0


In [13]:
turbine_14.shape

(1680, 5)

In [15]:
turbine_14.isnull().sum()

timestamp       0
turbine_id      0
wind_speed      0
power_kw        0
availability    0
dtype: int64

In [17]:
turbine_14.duplicated().sum()

np.int64(0)

In [11]:
dam_price_14 = pd.read_csv("../data_pack/dam_price_14day_sample.csv")
dam_price_14.head()

,timestamp,dam_price_inr_per_kwh
0,2026-03-01 00:00:00,4.255
1,2026-03-01 01:00:00,4.330
2,2026-03-01 02:00:00,4.186
3,2026-03-01 03:00:00,4.032
4,2026-03-01 04:00:00,4.141


In [14]:
dam_price_14.shape

(336, 2)

In [16]:
dam_price_14.isnull().sum()

timestamp                0
dam_price_inr_per_kwh    0
dtype: int64

In [18]:
dam_price_14.duplicated().sum()

np.int64(0)

For Part A, I will be working with two 14-day datasets: turbine_14 and dam_price_14.

turbine_14 contains hourly telemetry data for 5 turbines (T01–T05). For each hour, there are 5 rows, one corresponding to each turbine. The dataset has 1,680 rows and 5 columns, which matches the expected 14 × 24 × 5 = 1,680 records.

dam_price_14 contains the Day-Ahead Market (DAM) electricity price for each hour over the same 14-day period. It has 336 rows and 2 columns, matching 14 × 24 = 336 hourly records.
I checked both datasets and found no null values or duplicate rows. The expected hourly records are present throughout the 14-day period, so there are no missing observations in these Part A datasets.

In [19]:
turbine_90 = pd.read_csv("../data_pack/turbine_telemetry_90day.csv")
turbine_90.head()

,timestamp,turbine_id,wind_speed,power_kw,availability
0,2026-03-01 00:00:00,T01,7.11,198.3,1.0
1,2026-03-01 00:00:00,T02,5.42,25.6,1.0
2,2026-03-01 00:00:00,T03,8.31,393.2,1.0
3,2026-03-01 00:00:00,T04,4.82,11.3,1.0
4,2026-03-01 00:00:00,T05,4.24,27.3,1.0


In [22]:
turbine_90.shape 

(10725, 5)

In [27]:
turbine_90.isnull().sum()

timestamp       0
turbine_id      0
wind_speed      0
power_kw        0
availability    0
dtype: int64

In [25]:
turbine_90.duplicated().sum()

np.int64(0)

In [20]:
dam_price_90 = pd.read_csv("../data_pack/dam_price_90day.csv")
dam_price_90.head()

,timestamp,dam_price_inr_per_kwh
0,2026-03-01 00:00:00,4.255
1,2026-03-01 01:00:00,4.330
2,2026-03-01 02:00:00,4.186
3,2026-03-01 03:00:00,4.032
4,2026-03-01 04:00:00,4.141


In [23]:
dam_price_90.shape

(2160, 2)

In [24]:
dam_price_90.isnull().sum()

timestamp                0
dam_price_inr_per_kwh    0
dtype: int64

In [28]:
dam_price_90.duplicated().sum()

np.int64(0)

Part B — Dataset Understanding

For Part B, I will be using the 90-day turbine telemetry dataset (turbine_90) and the 90-day DAM price dataset (dam_price_90).

turbine_90 contains 10,725 rows and 5 columns. With 5 turbines recorded hourly over 90 days, we would expect 90 × 24 × 5 = 10,800 rows. Therefore, the dataset has 75 missing telemetry records.
dam_price_90 contains 2,160 rows and 2 columns, which matches the expected 90 × 24 = 2,160 hourly observations.
Both datasets contain no null values and no duplicate rows.

The missing telemetry records in turbine_90 therefore correspond to missing rows rather than null values, which is an important data-quality issue that needs to be investigated during the EDA. The provided README also explicitly states that missing telemetry rows should not be assumed to represent zero generation.